midend/optimizeSelectLabels.cpp

```diff
@@ -0,0 +1,428 @@
+                } else {
+                    if ((value >= previousLabel->rangeValue.first) &&
+                        (value <= previousLabel->rangeValue.second)) {
+                        alreadyAssigned = true;
+                        IR::Range* previousRange = getRange(previousLabel);
+                        ::warning(ErrorType::WARN_OVERLAPING,
+                                "Label value %1% is already assigned inside range:"
+                                " %2% .. %3%. Optimized code will not contain this label.",
+                                constant, previousRange->left, previousRange->right);
+                        break;
+                    }
+                }
+            }
+            if (!alreadyAssigned) {
+                vecOfLabels.push_back(label);
+                newVecCasses.push_back(currentExpression);
+            }
+        // Current label is range
+        } else if (expression->is<IR::Range>()) {
+            std::vector<SelectLabel*> elementsOfRange;
+            auto range = expression->to<IR::Range>();
+            auto leftConstant = range->left->to<IR::Constant>();
+            auto rightConstant = range->right->to<IR::Constant>();
+
+            if (leftConstant && rightConstant) {  // If labels of range are constants
+                auto left = leftConstant->value;
+                auto right = rightConstant->value;
+
+                std::pair<big_int, big_int> rangeValues(left, right);
+                SelectLabel* treeRoot = new SelectLabel(rangeValues, leftConstant->base,
+                                                        range->type, leftConstant->type);
+                bool insideRange = false;  // Define if range overlaping with previous label
+
+                // Go through all previous labels and check
+                // if some of them are overlaping with current range
+                for (SelectLabel* previousLabel : vecOfLabels) {
+                    // Constant value label
+                    if (previousLabel->type == SelectLabel::ConstantLabel) {
+                        if ((previousLabel->constValue >= rangeValues.first) &&
+                           (previousLabel->constValue <= rangeValues.second)) {
+                            IR::Constant* previousConstant =
+                               new IR::Constant(previousLabel->constValue, previousLabel->base);
+                            ::warning(ErrorType::WARN_OVERLAPING,
+                                "Element of range label %1% .. %2% is already assigned "
+                                "by constant label %3%. Optimization will parse this range.",
+                                range->left, range->right, previousConstant);
+                            insertValues(treeRoot, previousLabel->constValue);
+                            insideRange = true;
+                        }
+                    } else {
+                        if ((rangeValues.first >= previousLabel->rangeValue.first &&
```

Comment:
How about creating a utility function that can take two ranges and return an enum with values like "disjoint/inside/overlapleft/overlapright"? The you can reuse it everywhere and make the code much less prone to errors due to incorrect copy-and-paste.